Add winding factors analyzer to EM analyzers

dmnewman3 opened the pull request:
### **Winding Factor Analyzer**

When considering how to wind a stator, multiple factors need to be taken into account, such as the pole count, tooth sizing, and winding factor. The winding factor is particularly important as it directly affects the torque produced as the stator is excited. As such, an analyzer is being added to calculate winding factors based on the winding layout of a given stator. 

In the `b_field_outer_stator` analyzer within the `electromagnetic_analyzers` submodule of `mach_eval`, the winding factor is incorporated into the analyzer to map the b-field over one revolution. Currently, the code exists such that the winding factor(s) is manually recorded for each harmonic. While there is no issue with this, adding a separate winding factor analyzer will eliminate the need for any manual calculations, which can at times be quite complex. In the end, the `b_field_outer_stator` analyzer will incorporate the new `winding_factors` analyzer instead of manual calculations to enable easier integration of complex winding factors and an infinite amount of harmonics to be incorporated.

The inputs of the new `winding_factors` analyzer shall be the following:
- number of harmonics to be incorporated in the calculations
- winding layout of single- or double-layer windings for single phase
- angle of initial slot

For the following example, the inputs shall be the following:
- `number_of_harmonics` is `5`
- `winding_layout` is `np.array([[0,0,0,0,-1,-1,0,0,0,0,1,1],[0,0,-1,-1,0,0,0,0,1,1,0,0]])`
- `alpha_1` is `np.pi/12`

![Stator_Diagram](https://user-images.githubusercontent.com/112111913/224558870-ecb8659e-f4fd-4fff-bff9-8992c3cace50.svg)

The outputs of the new `winding_factors` analyzer shall be the following:
- complex winding factors for each harmonic being tested

The outputs of the stator pictured are:
- `kw_1` is `(0.4183 + 0.7244j)`
- `kw_2` is `(-3.053e-16 - 4.163e-17j)`
- `kw_3` is `(-8.327e-17 + 9.714e-17j)`
- `kw_4` is `(-3.469e-16 - 1.527e-16j)`
- `kw_5` is `(-0.1121 + 0.1941j)`

This new analyzer enables the complex winding factors to be calculated without the use of external code or calculations. With the addition of this analyzer, the implementation of an infinite number of winding factors becomes more realistic, as manual winding factor implementation is no longer required. This will make the visualization of current linkage diagrams and the full `b_field` using the `b_field_outer_stator` analyzer possible.

Bharat-Ramadas reviewed (commented):
Nice work @Aneesh-Jategaonkar
I would suggest the following changes:
1. Change the class docustring to reflect the details related to your cross-section
2.  Include a markdown document that will enable viewing the cross-section within Git. This [link](https://github.com/Severson-Group/eMach/tree/add-python-support/python/emach/model_obj/cross_sects/notched_rectangle) illustrates what I mean by this comment.

elsevers reviewed (changes requested):
Thanks Aneesh. However, this PR seems to be missing the `README.md` file necessary to be compliant with eMach cross sections. The image of the cross-section also is not using the naming convention.

Please review eMach's guidelines for creating new cross-sections [here](https://github.com/Severson-Group/eMach/tree/develop/model_obj/crosssects).

Bharat-Ramadas reviewed (commented):
Looks good to be merged in! 
Only comment is that the fillet feature has not been incorporated into the cross-section. Although this isn't critical, it might be a good idea to revise the cross-section image to remove the fillets so that there is no confusion moving forward.

elsevers commented:
> Looks good to be merged in! Only comment is that the fillet feature has not been incorporated into the cross-section. Although this isn't critical, it might be a good idea to revise the cross-section image to remove the fillets so that there is no confusion moving forward.

Actually, the fillet feature is quite important because we are planning to do structural analysis on this cross-sect. Without being able to set fillets, we'll likely have stress concentrations. So please do add the fillet feature.

elsevers reviewed (changes requested):
- [x] When I change your example to use `P=5`, JMAG crashes --see the error below. Please fix.
<img width="730" alt="image" src="https://user-images.githubusercontent.com/7421348/146618207-6ca19174-e1e2-45a6-82b3-aba885161493.png">

- [x] Please make `p` be lower case.
- [ ] Please implement fillets

Bharat-Ramadas reviewed (approved):
Looks good!

elsevers reviewed (changes requested):
This is getting very close. 

- [x] Update dimensions. Please take a look at the [inner notched rotor cross-section](https://github.com/Severson-Group/eMach/tree/Add_synrel_rotor_cross-section/python/emach/model_obj/cross_sects/inner_notched_rotor) and make your dimensions reasonably consistent with this:
    1. Change your `r_sh` to be `r_ri`
    2. Instead of `r_ri` use a yoke thickness `d_ri`
    3. Instead of `r_ro` use a pole depth of `d_rp`
- [x] Add a pole pitch annotation or change `p = number of pole pairs` to be `p = number of teeth`
- [x] There is no reason that this cross-section is only a synchronous reluctance rotor. It could be a switched reluctance rotor. Or, someone could even put permanent magnets between the teeth and make it a consequent pole rotor. We also have been specific in our rotor/stator naming of specifying whether it is for inner or outer. For all these reasons, let's rename this cross-section to be `Inner Reluctance Rotor`.

Bharat-Ramadas reviewed (commented):
Looks great! Only recommendation I have is to change `p` to `Q_r` so that users don't confuse this variable for number of pole pairs (Q_r for rotor teeth). Also note that this will also have to be changed in the .svg file. Apart from that, this PR looks ready to be merged in! Nice job!

Aneesh-Jategaonkar commented:
> Looks great! Only recommendation I have is to change `p` to `Q_r` so that users don't confuse this variable for number of pole pairs (Q_r for rotor teeth). Also note that this will have to be changed in the .svg file. Apart from that, this PR looks ready to be merged in! Nice job!

Variable name p is changed to Q_r and the .svg file is also updated

Bharat-Ramadas reviewed (approved):
Looks good!

elsevers reviewed (approved):
Nice job @Aneesh-Jategaonkar

dmnewman3 committed:
add winding factors analyzer to EM analyzers

- pair with bfield analyzer soon
- calculate winding factors based on winding layout

elsevers commented on mach_eval/analyzers/electromagnetic/winding_factors.py:
```diff
@@ -0,0 +1,86 @@
+class WindingFactorsProblem:
+    """Problem class for winding factor analyzer
+    Attributes:
+        harmonics: number of harmonics to be included in calculations []
```
`harmonics:` I think you mean for this to be a *list* of harmonics to calculate winding factors at

elsevers commented on mach_eval/analyzers/electromagnetic/winding_factors.py:
```diff
@@ -0,0 +1,86 @@
+                        and ending with bottom layer, length will be number 
+                        of teeth and height will be number of layers, "+" 
+                        for conductor into page, 0 for no present conductors, 
+                        and "-" for conductor out of page [turns]
```
Do you mean `+` and `-`, or do you mean `1` and `-1`?

elsevers commented on mach_eval/analyzers/electromagnetic/winding_factors.py:
```diff
@@ -0,0 +1,86 @@
+                        of teeth and height will be number of layers, "+" 
+                        for conductor into page, 0 for no present conductors, 
+                        and "-" for conductor out of page [turns]
+        alpha_1: angle of first slot from +x axis [rad]
```
Specify that this is the angle **counterclockwise** from the x axis

elsevers commented on mach_eval/analyzers/electromagnetic/winding_factors.py:
```diff
@@ -0,0 +1,86 @@
+            harmonics: harmonics array definied in problem
+            winding_layout: winding layout defined in problem
+            alpha_1: angle of slot 1 defined in problem
+            
```
Instead of specifying the arguments of the problem class's initializer, specify the arguments of the `analyze()` function here (this should be just the problem class)

elsevers reviewed (changes requested):
Great start @dmnewman3! See my feedback in line here.

Can you attach a Python script to your pull request (not in the repository) that I can use to see how the results of your winding function analyzer can be fed into Bharat's Bfield analyzer to create the current linkage plot for a phase winding?

dmnewman3 committed:
change harmonics variable to better reflect application

dmnewman3 committed:
change verbiage of analyzer function to reflect use of problem class

dmnewman3 committed:
change use of harmonics_list variable

dmnewman3 committed:
change use of harmonics_list variable

dmnewman3 committed:
remove extra variable based on change in harmonics_list setup

dmnewman3 committed:
account for changes in harmonics_list variable used in winding factor calculations

dmnewman3 commented:
@elsevers & @Anvar-Khamitov,

I've made the changes requested and (I think) I've added the code file I am using to test the integration into Bharat's stator analyzer. All I've done here is change is k_w definition to use my new analyzer. I am sure that my winding factor calculations are correct but the resulting graph using the b field, delta, and mu_o doesn't look like your typical current linkage plot. 

Please let me know if you cannot see the commits as I'm not sure if the changes have gotten to this pull request.

I have attached the code I am using to test as a .txt file as a .py file is not support to be uploaded. If code is usually uploaded as something different please advise.

Thank you,

Dante

[kwcode.txt](https://github.com/Severson-Group/eMach/files/10058217/kwcode.txt)

dmnewman3 committed:
add new winding analyzer document

dmnewman3 committed:
add section in source file

dmnewman3 committed:
add title to analyzer module

dmnewman3 committed:
add winding_factors to tree

dmnewman3 committed:
expand tree

dmnewman3 committed:
Delete winding_factors.rst

dmnewman3 committed:
change wording missed before

dmnewman3 committed:
add .csv file needed in winding factors documentation

dmnewman3 committed:
Add new winding_factors .rst file documentation

- implement changes not used in previous file version
- previous file version was deleted intentionally

dmnewman3 committed:
add new images needed in documentation

dmnewman3 committed:
change structure of equations

dmnewman3 committed:
no message

dmnewman3 committed:
fix equations

dmnewman3 committed:
edit documentation

elsevers reviewed (changes requested):
Thanks @dmnewman3, this is moving in the right direction, however I ran into an issue (described below) and wasn't able to fully review it. Feedback:

- [ ] I am unable to run your example code: `stator_Bn_prob` is not defined in this line: `mmf_comp = stator_Bn_prob.mmf(zq, Nc, n, k_w, I_hat) * np.cos(n * alpha + np.pi/2)`
- [ ] In `4.1 Model Background`: The `Stator_Diagram.svg` conductors look odd. The dots and crosses are not centered. Can you look into what is going on here? If you need help, let me know and I can take a look at your latex code  ![image](https://user-images.githubusercontent.com/7421348/206046228-e07e8db3-be8d-47cd-8f5d-6aee0479c1d3.png)
- [ ] In `4.1 Model Background`: It isn't proper machines theory to be talking about each coil phasor as a pitch factor. In fact, I don't think you should talk about pitch factor at all. We want this to be general to both double and single layer windings and pitch factor gets confusing for single layer windings
- [ ] In `4.1 Model Background`: I think you should be using `Stator_Diagram.svg` to indicate angles of slots for your winding factor calculation equation
- [ ] A thought for you to generalize your code: what if your `analyze()` function instead returns a `winding_factor` object that has a `kw(harmonics)` function that returns an array of complex winding factors for the harmonics passed into it? This would mean you wouldn't pass `harmonics` into your problem statement. I think it would make your returned value more useful. We can discuss.

dmnewman3 committed:
edit verbiage of background based on Eric's feedback

- add statement relating equation to Stator_Diagram.svg

dmnewman3 commented:
Eric,

Regarding your comments:
1) see attached new file that is relevant to the most current iteration of code
2) see attached PDF, I don't know why transfer to .svg is causing this. I have emailed you the latex code as well
3) changed, please review and provide comments
4) changed in text, please review and provide comments
5) I will work on a copy of that and have it ready to review with you tomorrow, I don't think that is a bad idea to consider

[Stator_Diagram.pdf](https://github.com/Severson-Group/eMach/files/10187184/Stator_Diagram.pdf)
[kwcode.txt](https://github.com/Severson-Group/eMach/files/10187160/kwcode.txt)

elsevers commented:
> Eric,
> 
> Regarding your comments:
> 
> 1. see attached new file that is relevant to the most current iteration of code
> 2. see attached PDF, I don't know why transfer to .svg is causing this. I have emailed you the latex code as well
> 3. changed, please review and provide comments
> 4. changed in text, please review and provide comments
> 5. I will work on a copy of that and have it ready to review with you tomorrow, I don't think that is a bad idea to consider
> 
> [Stator_Diagram.pdf](https://github.com/Severson-Group/eMach/files/10187184/Stator_Diagram.pdf) [kwcode.txt](https://github.com/Severson-Group/eMach/files/10187160/kwcode.txt)

Thanks Dante. But this doesn't seem to fix the code in the example provided within the RST file.

dmnewman3 committed:
fix problems with stator_image

dmnewman3 committed:
change errors in documentation and add code to enable mmf function to work

dmnewman3 committed:
add new current linkage plot

dmnewman3 committed:
add new .csv file for results

dmnewman3 committed:
change wording to make concept more clear

dmnewman3 committed:
add new documentation in .rst file

dmnewman3 commented:
Eric,

Please see new round of commits. I have added a bit more documentation and made some amendments from what we discussed over the past couple of days. Please let me know your thoughts.

Dante

dmnewman3 committed:
add file for B field plot

dmnewman3 committed:
add new if statement for radial vs. tangential air gap components

dmnewman3 committed:
change function calls to incorporate air gap and radius for the purpose of calculating the tangential field

dmnewman3 committed:
add new documentation in rst field for updates made using b field analyzer

dmnewman3 committed:
update figures

dmnewman3 committed:
update current linkage calculations

elsevers reviewed (dismissed):
Thanks @dmnewman3! Could you please split this into two PRs: 
1. one PR that fixes the issue in the B field outer stator
2. one PR that adds your new analyzer + RST file

dmnewman3 committed:
Revert "add new if statement for radial vs. tangential air gap components"

This reverts commit 1fbb7e4e3b0051c36c50d85c4de35f74ac6605a4.

dmnewman3 committed:
Revert "change function calls to incorporate air gap and radius for the purpose of calculating the tangential field"

This reverts commit d9460ec49262a758ed79c59f28930db1818b9c93.

elsevers commented:
Thanks @dmnewman3! Can you please work on revising the PR description to explain what you are adding and why.

dmnewman3 committed:
Merge branch 'dmnewman3_bfieldfix' into dmnewman3_kwanalyzer

dmnewman3 committed:
Fix winding layout image and winding factors

dmnewman3 committed:
fix index file

dmnewman3 committed:
Fix index file

dmnewman3 committed:
Fix index file

dmnewman3 committed:
Fix source file

dmnewman3 committed:
Replace incorrect images

dmnewman3 committed:
Fix documentation

dmnewman3 committed:
Fix code in documentation

dmnewman3 committed:
Merge branch 'develop' into dmnewman3_kwanalyzer

dmnewman3 commented:
All,
This should be ready for another review. Please provide any feedback you have.

Anvar-Khamitov reviewed (changes requested):
@dmnewman3 Thanks. I have some feedback on the analyzer .rst document:
* Typos: `usign`, `harmonnics`, `harmoincs_list`, harmonics of `interset`, `currently` linkage.
* I think it would be nice to highlight the words like `bfield_outer_stator`, `harmonics_list`, `B Field Outer Stator Analyzer`, `"k_w"`, `"n"`, i.e. to make them look like:
![image](https://user-images.githubusercontent.com/42312168/225445653-0608c37d-190c-43e7-9576-797cb123fb91.png)
* This sentence seemed confusing: `They can be thought of as a proportion of the geometric vector sum of coil side phases in a phase winding over the algebraic sum`. Maybe you could use something similar to `...geometric vector sum of the phasors of a particular back-EMF harmonic induced in coil sides in a phase winding...`.
* Line 96: you used `alpha_1 = 0`. To get the same winding factor results that you showed in the document, I think it should be set to `alpha_1 = np.pi/12`.
* Current linkage plot: 1) line 195: you used `alpha` in radians when making a plot, while the plot in the document is in deg, 2) line 136: to get the same current linkage plot as in the document, I had to change `winding_layout` in line 136 to the original one in line 95.

Otherwise, everything else looks good to me.

dmnewman3 committed:
Fix typos

dmnewman3 committed:
Edit sentence in .rst file to avoid confusion

dmnewman3 committed:
Change alpha_1 from 0 to np.pi/12

dmnewman3 committed:
Replace winding layout and change radians to degrees to match plot included in PR

dmnewman3 committed:
Highlight variables in .rst file

dmnewman3 commented:
Sounds good @Anvar-Khamitov, thanks for catching those!

@elsevers let me know what you think!

dmnewman3 committed:
reverse bfieldfixes

dmnewman3 committed:
Revert "reverse bfieldfixes"

This reverts commit 96338534e15f13178bb605849748a23d8c817ba4.

Anvar-Khamitov commented on mach_eval/analyzers/electromagnetic/bfield_outer_stator.py:
```diff
@@ -50,7 +50,8 @@ def mmf(self, m, zq, N, n, k_w, I_hat):
         """
-        mmf = m / np.pi * zq * N * k_w * I_hat / n
+        
+        mmf = 2 / np.pi * zq * Nc * k_w * I_hat / n
```
@dmnewman3 I'm wondering what is the reason for this change in the mmf expression? Based on the comment [here](https://github.com/Severson-Group/eMach/pull/229#issuecomment-1480461065), the cumulative mmf from all phases would be `mmf = m / np.pi * zq * N * k_w * I_hat / n`.

dmnewman3 replied:
@Anvar-Khamitov, that's a mistake on my part, I just fixed this. This should also go away once Issue #232 is closed, which will happen before this PR is advance and approved.

Anvar-Khamitov replied:
@dmnewman3, I suppose your field and current linkage plots will have to be changed. Otherwise, everything looks good to me.

dmnewman3 replied:
Sounds good @Anvar-Khamitov, I just replaced the images and slightly altered the .rst file to produce the images seen in the html.

Please let me know what you think and if you need me to make any changes.

Anvar-Khamitov replied:
@dmnewman3 looks good! But it seems like this branch has some merge conflicts.

dmnewman3 committed:
fix mmf expression

dmnewman3 committed:
Merge b_field_fix

dmnewman3 committed:
Update winding_factors.rst to account for edits made in b_field_fix branch recently merged with develop

dmnewman3 committed:
Replace .png images with .svg images

dmnewman3 committed:
Replace images

dmnewman3 committed:
Replace images

dmnewman3 committed:
Revert "Replace images"

This reverts commit f369fa3287fe29d388b94dd90cc7c3693e5eeee5.

dmnewman3 committed:
Revert "Revert "Replace images""

This reverts commit af9dbc0c9e6dcfac0035c33847508e9c28a6f369.

dmnewman3 committed:
Merge branch 'develop' into dmnewman3_kwanalyzer

dmnewman3 committed:
Replace problematic images

npetersen2 reviewed (dismissed):
@dmnewman3 this looks great to me! Awesome to have this analyzer :)

I did not check the math, but overall, the code / plots / docs look good to me.

My only comment is: your python code is rather long/split up for no real reason (that I can see). I am not sure if the `calculations()` method is needed, can't you just put the calcs in the analyze method?  In the end, this doesn't matter, but a bit odd looking to me.

npetersen2 commented:
Also, in PRs, usually the PR author "resolves" all open comments after they have been resolved, but I see many open comments in the conversation above. Are these done? If so, can you resolve them? thanks

npetersen2 commented on mach_eval/analyzers/electromagnetic/bfield_outer_stator.py:
```diff
@@ -36,7 +36,7 @@ def __init__(self, m, zq, Nc, k_w, I_hat, n, delta_e, r_si, r_rfe, alpha_so):
         self.alpha_so = alpha_so
 
-    def mmf(self, m, zq, N, n, k_w, I_hat):
+    def mmf(self, m, zq, Nc, n, k_w, I_hat):
```
Is `Nc` the new nomenclature here? why the change from `N` to `Nc`?

npetersen2 replied:
The comment for `args` says "N"

dmnewman3 committed:
Correct erroneous variable name Nc to N

dmnewman3 commented:
@elsevers @Anvar-Khamitov @npetersen2 @wchan29 
This is ready to be reviewed for the final time and pulled if approved.

npetersen2 reviewed (dismissed):
Looks good to me! Thanks for adding this

Anvar-Khamitov reviewed (dismissed):
Looks good to me!

elsevers committed:
edit text

elsevers committed:
edit winding factors

elsevers reviewed (approved):
Nice job @dmnewman3! I finally managed to get my review in here and I like what you've done with this. I made a few language edits. Go ahead and merge when you are happy with it.

dmnewman3 commented:
> Nice job @dmnewman3! I finally managed to get my review in here and I like what you've done with this. I made a few language edits. Go ahead and merge when you are happy with it.

Sounds good @elsevers, thank you
